File: 104.Maximum-Depth-of-Binary-Tree/tree.c
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

int maxDepth(struct TreeNode *root) {
    if (!root) {
        return 0;
    }

    // O(n)
    int l = maxDepth(root->left);
    int r = maxDepth(root->right);
    return 1 + (l > r ? l : r);
}

File: 226.Invert-Binary-Tree/tree.rs
use std::{cell::RefCell, rc::Rc};

impl Solution {
    pub fn invert_tree(root: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        root.map(|node| {
            {
                let mut borr = node.borrow_mut();
                let left = borr.left.take();
                let right = borr.right.take();

                // O(n)
                borr.right = Self::invert_tree(left);
                borr.left = Self::invert_tree(right);
            }

            node
        })
    }
}
